libsmtutil/CVC5Interface.cpp

```diff
@@ -0,0 +1,319 @@
+/*
+	This file is part of solidity.
+
+	solidity is free software: you can redistribute it and/or modify
+	it under the terms of the GNU General Public License as published by
+	the Free Software Foundation, either version 3 of the License, or
+	(at your option) any later version.
+
+	solidity is distributed in the hope that it will be useful,
+	but WITHOUT ANY WARRANTY; without even the implied warranty of
+	MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+	GNU General Public License for more details.
+
+	You should have received a copy of the GNU General Public License
+	along with solidity.  If not, see <http://www.gnu.org/licenses/>.
+*/
+// SPDX-License-Identifier: GPL-3.0
+
+#include <libsmtutil/CVC5Interface.h>
+
+#include <libsolutil/CommonData.h>
+#include <libsolutil/CommonIO.h>
+#include <libsolutil/Exceptions.h>
+
+using namespace std;
+using namespace solidity;
+using namespace solidity::util;
+using namespace solidity::smtutil;
+
+CVC5Interface::CVC5Interface(optional<unsigned> _queryTimeout):
+	SolverInterface(_queryTimeout)
+{
+	reset();
+}
+
+void CVC5Interface::reset()
+{
+	m_variables.clear();
+	m_solver.resetAssertions();
+	m_solver.setOption("produce-models", "true");
+	if (m_queryTimeout)
+		m_solver.setOption("tlimit-per", std::to_string(*m_queryTimeout));
+	else
+		m_solver.setOption("rlimit", std::to_string(resourceLimit));
+}
+
+void CVC5Interface::push()
+{
+	m_solver.push();
+}
+
+void CVC5Interface::pop()
+{
+	m_solver.pop();
+}
+
+void CVC5Interface::declareVariable(string const& _name, SortPointer const& _sort)
+{
+	smtAssert(_sort, "");
+	m_variables[_name] = m_solver.mkConst(CVC5Sort(*_sort), _name);
+}
+
+void CVC5Interface::addAssertion(Expression const& _expr)
+{
+	try
+	{
+		m_solver.assertFormula(toCVC5Expr(_expr));
+	}
+	catch (cvc5::CVC5ApiException const& _e)
```

Comment:
Before we were catching 4 exceptions, are you sure this is enough?

Reply:
According to the documentation, there are only two classes of API exceptions: `CVC5ApiException` and  `CVC5ApiRecoverableException`, with the second being a subclass of the former.
The difference is that the solver is in a consistent state on the latter case.
My opinion is that we do not need to distinguish the second case.

https://cvc5.github.io/docs/cvc5-1.0.2/api/cpp/exceptions.html